metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/evaluator/BaselineRuleEvaluator.java

```diff
@@ -0,0 +1,194 @@
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.EventRecord;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.MetricAnomalyNotificationEvent;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.NotificationEvent;
+import org.hypertrace.alert.engine.metric.anomaly.datamodel.ViolationSummary;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.hypertrace.core.query.service.api.ResultSetChunk;
+import org.hypertrace.core.query.service.api.Row;
+import org.hypertrace.gateway.service.v1.baseline.Baseline;
+import org.hypertrace.gateway.service.v1.common.Value;
+import org.hypertrace.gateway.service.v1.common.ValueType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BaselineRuleEvaluator {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BaselineRuleEvaluator.class);
+
+  private final QueryRequestHandler queryRequestHandler;
+
+  public BaselineRuleEvaluator(QueryRequestHandler queryRequestHandler) {
+    this.queryRequestHandler = queryRequestHandler;
+  }
+
+  Optional<NotificationEvent> evaluateRule(
+      MetricAnomalyEventCondition metricAnomalyEventCondition, AlertTask alertTask)
+      throws IOException {
+    DynamicThresholdCondition dynamicThresholdCondition =
+        metricAnomalyEventCondition
+            .getViolationConditionList()
+            .get(0)
+            .getDynamicThresholdCondition();
+
+    Duration duration = java.time.Duration.parse(dynamicThresholdCondition.getBaselineDuration());
+    long durationMillis = duration.toMillis();
+
+    // this query will fetch metric data, which will used for baseline calculation and evaluation
+    QueryRequest queryRequest =
+        queryRequestHandler.getQueryRequest(
+            metricAnomalyEventCondition,
+            alertTask.getTenantId(),
+            alertTask.getCurrentExecutionTime() - durationMillis,
+            alertTask.getCurrentExecutionTime());
+
+    LOGGER.debug("Rule id {}, Query request {}", alertTask.getEventConditionId(), queryRequest);
+
+    Iterator<ResultSetChunk> iterator =
+        queryRequestHandler.executeQuery(
+            Map.of(TENANT_ID_KEY, alertTask.getTenantId()), queryRequest);
+
+    List<Double> metricValuesForBaseline = new ArrayList<>();
+    List<Double> metricValuesForEvaluation = new ArrayList<>();
+
+    iterator.forEachRemaining(
+        chunk -> {
+          for (Row row : chunk.getRowList()) {
+            if (row.getColumnCount() >= 2
+                && row.getColumn(1).getValueType()
+                    == org.hypertrace.core.query.service.api.ValueType.STRING) {
+              double value = Double.parseDouble(row.getColumn(1).getString());
+              metricValuesForBaseline.add(value);
+              if (Long.parseLong(row.getColumn(0).getString())
+                  >= alertTask.getLastExecutionTime()) {
+                metricValuesForEvaluation.add(value);
+              }
```

Comment:
Shouldn't this will come in else block? This one is adding current values as well in baseline calculation. This is not the case currently in `gateway`.
```
if (Long.parseLong(row.getColumn(0).getString())
                  >= alertTask.getLastExecutionTime()) {
                metricValuesForEvaluation.add(value);
              } else {
metricValuesForBaseline.add(value);
}
```

Reply:
this line `metricValuesForEvaluation.add(value);` doesn't have anything to do with gateway logic of baseline calculation

Reply:
oh, so `CurrentExection - baselineTime` this is the duration considered for baseline duration

Reply:
This is what I was saying that, we should not include the current min. So, it should be
`LastExecutionTime - baselineTime` . As far as I know, that's how it was done in the `gateway` service, as we are using the same function.

Reply:
why should it be `LastExecutionTime - baselineTime` ? 
the concept of execution time is only related to rule evaluation right, so how come there is equivalent logic present on gateway side?

Reply:
The last execution time is the previous min, right?

Reply:
yes